src/redisearch_rs/rqe_iterators/src/metric.rs

```diff
@@ -7,28 +7,58 @@
  * GNU Affero General Public License v3 (AGPLv3).
 */
 
-//! Metric iterator implementation
-
 use crate::{RQEIterator, RQEIteratorError, SkipToOutcome, id_list::IdList};
-use ffi::{RSYieldableMetric, array_ensure_append_n_func, t_docId};
+use ffi::{RLookupKey, RLookupKeyHandle, RSYieldableMetric, array_ensure_append_n_func, t_docId};
 use inverted_index::{RSIndexResult, ResultMetrics_Reset_func};
 use value::{RSValueFFI, RSValueTrait};
 
-// enum indicating the type of metric. currently only vector distance is supported.
+/// The different types of metrics.
+/// At the moment, only vector distance is supported.
+#[repr(C)]
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
 pub enum MetricType {
     VectorDistance,
 }
 
-/// An iterator that yields all ids within a given range, from 1 to max id (inclusive) in an index.
-pub struct Metric<'index> {
-    base: IdList<'index>,
+/// An iterator that yields document ids alongside a metric value (e.g. a score or a distance),
+/// sorted by document id.
+pub type MetricIteratorSortedById<'index> = MetricIterator<'index, true>;
+/// An iterator that yields document ids alongside a metric value (e.g. a score or a distance),
+/// sorted by metric value.
+pub type MetricIteratorSortedByScore<'index> = MetricIterator<'index, false>;
+
+/// An iterator that yields document ids alongside a metric value (e.g. a score or a distance).
+/// The iterator can be sorted by document id or by metric value,
+/// but the choice is made at compile time.
+pub struct MetricIterator<'index, const SORTED_BY_ID: bool> {
+    base: IdList<'index, SORTED_BY_ID>,
     metric_data: Vec<f64>,
     #[allow(dead_code)]
     type_: MetricType,
+    own_key: *mut RLookupKey,
+    /// # Invariants
+    ///
+    /// The handle is either:
+    ///
+    /// - A null pointer, indicating that the iterator is not associated with a key.
+    /// - A valid pointer to a [`RLookupKeyHandle`] instance.
+    key_handle: *mut RLookupKeyHandle,
```

Comment:
Is this addition related directly to the ability to create an unsorted variant of a metric iterator?